### PR TITLE
Payment-Stuff

### PR DESCRIPTION
**Pull request by lckr**
- payment-package

### CODE REVIEW

**Comment by LeanderK** [comment]
i can review it

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/payment/PaymentCalculator.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+        int paymentBase = experiment.getPaymentBase();
+        int paymentAnswer = experiment.getPaymentAnswer();
+        int paymentRating = experiment.getPaymentRating();
+
+        if(paymentAnswer < 0 || paymentBase < 0 || paymentRating < 0){
+            throw new IllegalArgumentException("Error in "+this.getClass()+"! Payment-values must not be less than zero!");
+        }
+
+        //Gets a all answers of a worker above the specified threshold
+        Map<WorkerRecord,Set<AnswerRecord>> workerAnswerSet = ops.getGoodAnswersOfExperiment(experiment.getId(), 0); //TODO params
```
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/payment/PaymentCalculator.java]: whats with this TODO?

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/payment/PaymentCalculator.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+            if(!map.containsKey(worker)) {
+                map.put(worker, 0);
+            }
+
+            for (int i = 0; i < entry.getValue().size(); i++) {
+                map.put(worker, map.get(worker) + paymentAnswer);
+            }
+        }
+
+        Map<WorkerRecord,Set<RatingRecord>> workerRatingSet = ops.getGoodRatingsOfExperiment(experiment.getId(), 0); //TODO ratings params
```
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/payment/PaymentCalculator.java]: comment above

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/payment/PaymentCalculator.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+ * Created by lucaskrauss at 28.01.2016
+ *
+ * The PaymentCalculator calculates the payment of all workers of an experiment.
+ * The payment of a worker depends on the number of good (above the specified threshold) answers and good ratings.
+ * If the experiment's base payment is greater than zero the worker's payment is increased by that value.
+ */
+public class PaymentCalculator {
+
+
+    private AnswerRatingOperations ops;
```
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/payment/PaymentCalculator.java]: final?

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/payment/PaymentCalculator.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+        int paymentBase = experiment.getPaymentBase();
+        int paymentAnswer = experiment.getPaymentAnswer();
+        int paymentRating = experiment.getPaymentRating();
+
+        if(paymentAnswer < 0 || paymentBase < 0 || paymentRating < 0){
+            throw new IllegalArgumentException("Error in "+this.getClass()+"! Payment-values must not be less than zero!");
+        }
+
+        //Gets a all answers of a worker above the specified threshold
+        Map<WorkerRecord, List<AnswerRecord>> workerAnswerSet = ops.getGoodAnswersOfExperiment(experiment.getId(), 0); //TODO params
```
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/payment/PaymentCalculator.java]: TODO? did you forget something?

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/payment/PaymentCalculator.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+            if(!map.containsKey(worker)) {
+                map.put(worker, 0);
+            }
+
+            for (int i = 0; i < entry.getValue().size(); i++) {
+                map.put(worker, map.get(worker) + paymentAnswer);
+            }
+        }
+
+        Map<WorkerRecord, List<RatingRecord>> workerRatingSet = ops.getGoodRatingsOfExperiment(experiment.getId(), 0); //TODO ratings params
```
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/payment/PaymentCalculator.java]: above

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/payment/PaymentCalculator.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+        for (Map.Entry<WorkerRecord, List<AnswerRecord>> entry : workerAnswerSet.entrySet()) {
+            Worker worker = WorkerTransformer.toProto(entry.getKey());
+
+            if(!map.containsKey(worker)) {
+                map.put(worker, 0);
+            }
+
+            for (int i = 0; i < entry.getValue().size(); i++) {
+                map.put(worker, map.get(worker) + paymentAnswer);
+            }
```
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/payment/PaymentCalculator.java]: i think this is very complicated. entry.getValue().size*paymentAnswer would do the job. I can give you the size directly. I don't have to give you all the answers.

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/payment/PaymentCalculator.java [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+        for (Map.Entry<WorkerRecord, List<RatingRecord>> entry : workerRatingSet.entrySet()) {
+            Worker worker = WorkerTransformer.toProto(entry.getKey());
+
+            if(!map.containsKey(worker)) {
+                map.put(worker, 0);
+            }
+
+            for (int i = 0; i < entry.getValue().size(); i++) {
+                map.put(worker, map.get(worker) + paymentRating);
+            }
```
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/payment/PaymentCalculator.java]: see above

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/payment/PaymentCalculator.java (hunk 4 above)
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/payment/PaymentCalculator.java]: i would add the parameters before merging with master. Just ask me how they are implemented in the database then we can work this out

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/payment/PaymentCalculator.java [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+        }
+
+
+        //Now all workers participating in the experiment are present in the map and the base-payment can be added to their salary
+        if(paymentBase > 0) {
+            for (Map.Entry<Worker, Integer> entry : map.entrySet()) {
+                entry.setValue(entry.getValue() + paymentBase);
+            }
+        }
```
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/payment/PaymentCalculator.java]: i think this if statement is a bit unnecessary.

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/payment/PaymentCalculator.java [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+            for (int i = 0; i < entry.getValue().size(); i++) {
+                map.put(worker, map.get(worker) + paymentRating);
+            }
+        }
+
+
+        //Now all workers participating in the experiment are present in the map and the base-payment can be added to their salary
+        if(paymentBase > 0) {
+            for (Map.Entry<Worker, Integer> entry : map.entrySet()) {
+                entry.setValue(entry.getValue() + paymentBase);
```
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/payment/PaymentCalculator.java]: entry.setValue is optional, i would resort to the good old map.put

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/payment/PaymentDispatcher.java [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+     * The PaymentDispatcher checks if the state of the experiment has changed from STOPPING to STOPPED.
+     * In that case the payment is calculated (via PaymentCalculator) and dispatched to the PlatformManager.
+     *
+     * @param experimentChangeEvent representation of the changed experiment before and after the change
+     */
+    @Override
+    public void onNext(ChangeEvent<Experiment> experimentChangeEvent) {
+
+        if (experimentChangeEvent.getNeww().getState() == Experiment.State.STOPPED
+                && experimentChangeEvent.getOld().getState() == Experiment.State.CREATIVE_STOPPED) {
```
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/payment/PaymentDispatcher.java:66]: `CREATIVE_STOPPED` does not mean the experiment stopped, just the creative-part stopped

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/payment/PaymentDispatcher.java [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+
+    /**
+     * Gets a map of workers and their corresponding payment via the PaymentCalculator.
+     * Dispatches the salary of the workers via the PlatformManager
+     *
+     * @param exp the finished experiment
+     */
+    private void dispatchPayment(Experiment exp) {
+        Map<Worker, Integer> map;
+        map = paymentCalc.estimatePayment(exp);
```
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/payment/PaymentDispatcher.java]: weird initialisation

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/payment/PaymentDispatcher.java [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+
+        map.forEach((Worker, Integer) -> platformManager.payWorker("", Worker, Integer));
```
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/payment/PaymentDispatcher.java]: why empty string? this does not seem right...

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/payment/PaymentDispatcher.java [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+ * The PaymentDispatcher has a subscription to an EXPERIMENT_CHANGE-Observable.
+ * On experiment-end, the PaymentDispatcher collects the payments per worker via the PaymentCalculator
+ * and uses the results of that calculation to dispatch the payments to the PlatformManager.
+ */
+public class PaymentDispatcher implements Observer<ChangeEvent<Experiment>> {
+
+
+    private Observable<Event<ChangeEvent<Experiment>>> observable = EventManager.EXPERIMENT_CHANGE.getObservable();
+    private PlatformManager platformManager;
+    private PaymentCalculator paymentCalc;
```
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/payment/PaymentDispatcher.java]: final?

File: src/test/java/edu/kit/ipd/crowdcontrol/objectservice/payment/PaymentCalculatorTest.java [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+    @Test (expected = IllegalArgumentException.class)
+    public void testFalsePaymentArgs(){
+        exp = Experiment.newBuilder().setPaymentBase(-1).setPaymentAnswer(4).setPaymentRating(4).build();
+        calculator.estimatePayment(exp);
+
+    }
+
+
+    private List<Worker> sortWorkerMap(Map<Worker,Integer> map){
+        List<Worker> sortedWorkers = new LinkedList<>(map.keySet());
```
Comment (LeanderK) [inline src/test/java/edu/kit/ipd/crowdcontrol/objectservice/payment/PaymentCalculatorTest.java]: just a general advice...don't use LinkedLists in java, they only provide benefits when adding to head or tail of huge lists. But then you should have used Queues.

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/payment/PaymentDispatcher.java (hunk 10 above)
Comment (lckr) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/payment/PaymentDispatcher.java:66]: I'm testing if the state changed from `CREATIVE_STOPPED` to `STOPPED` , in that case the experiment should have stopped

**Comment by LeanderK** [comment]
i will add the parameters, then i (or you) can open the pull request again.

**Pull request closed without merging**